ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -576,31 +581,11 @@ private boolean isExpired(IndexValue value) {
   FindInfo findEntriesSince(FindToken token, long maxTotalSizeOfEntries) throws StoreException {
     long startTimeInMs = time.milliseconds();
     try {
-      boolean tokenWasReset = false;
+      boolean tokenWasReset;
       Offset logEndOffsetBeforeFind = log.getEndOffset();
-      StoreFindToken storeToken = (StoreFindToken) token;
-      // validate token
-      if (storeToken.getSessionId() == null || storeToken.getSessionId().compareTo(sessionId) != 0) {
-        // the session has changed. check if we had an unclean shutdown on startup
-        if (!cleanShutdown) {
-          // if we had an unclean shutdown and the token offset is larger than the logEndOffsetOnStartup
-          // we reset the token to logEndOffsetOnStartup
-          if (!storeToken.getType().equals(StoreFindToken.Type.Uninitialized)
-              && storeToken.getOffset().compareTo(logEndOffsetOnStartup) > 0) {
-            logger.info("Index : " + dataDir + " resetting offset after not clean shutdown " + logEndOffsetOnStartup
-                + " before offset " + storeToken.getOffset());
-            storeToken = new StoreFindToken(logEndOffsetOnStartup, sessionId);
-            tokenWasReset = true;
-          }
-        } else if (!storeToken.getType().equals(StoreFindToken.Type.Uninitialized)
-            && storeToken.getOffset().compareTo(logEndOffsetOnStartup) > 0) {
-          logger.error(
-              "Index : " + dataDir + " invalid token. Provided offset is outside the log range after clean shutdown");
-          // if the shutdown was clean, the offset should always be lesser or equal to the logEndOffsetOnStartup
-          throw new IllegalArgumentException(
-              "Invalid token. Provided offset is outside the log range after clean shutdown");
-        }
-      }
+      AtomicReference<StoreFindToken> storeFindTokenAtomicReference = new AtomicReference<>((StoreFindToken) token);
```

Comment:
Why is a `StoreToken` that is externally visible never "inclusive" though? In fact, I even think that there is a bug here. Consider this scenario:
1. Incoming token is a journal based token with an offset higher than the `logEndOffsetOnStartup`, say.
2. The store resets it to `logEndOffsetOnStartup` - but marks the token as inclusive so that it does not miss a new write that may go in at `logEndOffsetOnStartup`.
3. Let us say no write has gone into the store yet since startup, and the store ends up returning this token as-is (I see that this can happen in one code path for sure; might not happen in all code paths though the code does not seem to explicitly ensure against the token getting returned as-is).
4. The next time the caller makes a request and passes in this token, the knowledge that the entries returned should include the entry at this offset is lost, and that entry might not get replicated at all.

Am I missing something?

The point I am trying to make is that the token represents a specific position and the "inclusive" flag is *necessary* to representing that position correctly. It should therefore be inherent to the token, and should be part of what gets sent across the wire and back, and persisted, and so on, since `findEntriesSince()` method is stateless.